core/src/test/java/io/grpc/DeadlineTest.java

```diff
@@ -47,60 +45,62 @@
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
+import java.util.Random;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
 
 /**
  * Tests for {@link Context}.
  */
 @RunWith(JUnit4.class)
 public class DeadlineTest {
+  private FakeTicker ticker = new FakeTicker();
 
-  // Allowed inaccuracy when comparing the remaining time of a deadline.
-  private final long maxDelta = TimeUnit.MILLISECONDS.toNanos(20);
+  @Test
+  public void defaultTickerIsSystemTicker() {
+    Deadline d = Deadline.after(0, TimeUnit.SECONDS);
+    ticker.reset(System.nanoTime());
+    Deadline reference = Deadline.after(0, TimeUnit.SECONDS, ticker);
+    // Allow inaccuracy to account for system time advancing during test.
+    assertAbout(deadline()).that(d).isWithin(20, TimeUnit.MILLISECONDS).of(reference);
+  }
 
   @Test
   public void immediateDeadlineIsExpired() {
-    Deadline deadline = Deadline.after(0, TimeUnit.SECONDS);
+    Deadline deadline = Deadline.after(0, TimeUnit.SECONDS, ticker);
     assertTrue(deadline.isExpired());
   }
 
   @Test
   public void shortDeadlineEventuallyExpires() throws Exception {
-    Deadline d = Deadline.after(100, TimeUnit.MILLISECONDS);
+    Deadline d = Deadline.after(100, TimeUnit.MILLISECONDS, ticker);
     assertTrue(d.timeRemaining(TimeUnit.NANOSECONDS) > 0);
     assertFalse(d.isExpired());
-    Thread.sleep(101);
+    ticker.increment(101, TimeUnit.MILLISECONDS);
 
     assertTrue(d.isExpired());
-    assertFalse(d.timeRemaining(TimeUnit.NANOSECONDS) > 0);
-    assertAbout(deadline()).that(d).isWithin(maxDelta, NANOSECONDS).of(Deadline.after(0, SECONDS));
+    assertEquals(-1, d.timeRemaining(TimeUnit.MILLISECONDS));
   }
 
   @Test
   public void deadlineMatchesLongValue() {
-    long minutes = Deadline.after(10, TimeUnit.MINUTES).timeRemaining(TimeUnit.MINUTES);
-
-    assertTrue(minutes + " != " + 10, Math.abs(minutes - 10) <= 1);
+    assertEquals(10, Deadline.after(10, TimeUnit.MINUTES, ticker).timeRemaining(TimeUnit.MINUTES));
   }
 
   @Test
   public void pastDeadlineIsExpired() {
-    Deadline d = Deadline.after(-1, TimeUnit.SECONDS);
+    Deadline d = Deadline.after(-1, TimeUnit.SECONDS, ticker);
     assertTrue(d.isExpired());
-
-    assertAbout(deadline()).that(d).isWithin(maxDelta, NANOSECONDS).of(Deadline.after(-1, SECONDS));
+    assertEquals(-1000, d.timeRemaining(TimeUnit.MILLISECONDS));
   }
 
   @Test
   public void deadlineDoesNotOverflowOrUnderflow() {
-    Deadline after = Deadline.after(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
+    Deadline after = Deadline.after(Long.MAX_VALUE, TimeUnit.NANOSECONDS, ticker);
```

Comment:
Since the random is removed, shall we reset the ticker at the beginning to a non-zero value?
A naive implementation that doesn't anything to prevent overflow would also pass the test if the ticker is at zero.